game.py

```diff
@@ -0,0 +1,908 @@
+from math import inf
+from copy import copy, deepcopy
+import random
+import pickle
+import time
+
+class GameState():
+    
+    def __init__(self, board, player_next, connect_number, number_of_moves, moves_list = []):
+        self._board = board
+        self._player_next = player_next
+        self._connect_number = connect_number
+        self._number_of_moves = number_of_moves
+        self._moves_list = moves_list
+        # self._last_move = last_move
+        # self._previous_game_state = previous_game_state
+
+    def _copy(self):
+        board = deepcopy(self._board)
+        player_next = deepcopy(self._player_next)
+        connect_number = deepcopy(self._connect_number)
+        number_of_moves = deepcopy(self._number_of_moves)
+        moves_list = deepcopy(self._moves_list)
+        # previous_game_state = deepcopy(self)
+        return GameState(board, player_next, connect_number, number_of_moves, moves_list)
+                
+    def get_board(self):
+        return self._board
+    
+    def get_player_next(self):
+        return self._player_next
+
+    def get_last_move(self):
+        return self._moves_list[-1]
+    
+    def get_connect_number(self):
+        return self._connect_number
+
+    def get_number_of_moves(self):
+        return self._number_of_moves
+    
+    def make_move(self, move, player_next = True):
+               
+        self._board[move[0]][move[1]] = self._player_next
+        self._player_next = - self._player_next
+        self._moves_list.append(move)
+        self._number_of_moves = self._number_of_moves + 1
+
+        # return game_state
+        
+    def unmake_move(self):
+        move = self.get_last_move()
+        self._board[move[0]][move[1]] = None
+        self._player_next = - self._player_next
+        self._number_of_moves = self._number_of_moves - 1
+        self._moves_list = self._moves_list[: -1]
+        
+        # return game_state
+
+    def get_next_moves(self):    
+    
+        global columns
+        global rows
+        global column_order
+                
+        moves = []
+        any_moves = False
+
+        for column in range(columns):
+            if any_moves:
+                break
+            for row in range(rows):
+                if self._board[row][column_order[column]] == None:
+                    moves.append ((row, column_order[column]))
+                    break
+                
+        return moves
+   
+    def game_end(self):
+
+# This determines whether the game has ended after the last move and whether it is a win, lose
+# or draw (if the board is full)
+
+        if self.get_number_of_moves() == 0:
+            return False, False        
+        player = - self.get_player_next()
+        board = self.get_board()

```

Comment:
This is a pattern of grabbing the board just at the beginning of the function (although I do see another `self.get_board()` has snuck in right at the end of the function).

I think this is preferable with how things are setup right now, but with a different class-hierarchy could maybe but setup differently.
* Right now, I think it's a nice indicator to the reader that you don't expect the board-state to be changing under you while executing other methods. It's clear with a quick scan of the rest of the function that you just read from board, rather than assign to it

On the flip side, if this was part of a larger program I might consider things like:
* Having `self.get_board()` return a copy of the board rather than internal state to this class. Currently, we're allowing a user outside of the class to grab the state and then potentially mutate it.
* Leaning further into "proper OOP", (which would generally be overkill for your program), this method might not belong on this class (I can discuss more elsewhere on what should/shouldn't). If it's no longer on the class, then code like `board[next_position[0]][next_position[1]]` for me is bleeding a lot of internal knowledge about how boards and positions are represented in memory, which I'd rather not do. Think about moving this method off this class, but adding extra methods (and maybe classes) like so:

```
@dataclass(frozen=True)
class BoardPosition:
    column: int
    row: int

class Tile(Enum):
    PLAYER_ONE
    PLAYER_TWO
    EMPTY

class Board:
    ...
    def get_tile(self, position: BoardPosition) -> Tile:
        pass

```